---
title: "Finish Aligning and Generate Read Counts"
author: "Noelle Wheeler"
date: "2023-03-22"
output: html_document
---

Github url: https://github.com/nwh84/angsd-project
All bash scripts are in github repo

### Align samples
```{}
$ sbatch align_all.sh
```

### Run multiQC on flagstat and fastqc
```{}
sbatch run_fastQC.sh $SCRATCH_DIR/fastqc_out
```
[multiQC-output](https://htmlpreview.github.io/?https://github.com/nwh84/angsd-project/blob/main/QC/multiqc.ouput.html)

Output .html file in github
This shows the summary of the samtools output and the multiqc output. We can see that SRR8368044_BP1 has far more reads mapped than the other samples. However, the total sequences count for the paired end reads looks similar to the other sequences. The high number of mapped reads may be due to a high level of PCR duplication or because the sample has many repetitive regions. When we look at fastQC results, we don't see high levels of duplicate reads for this sample, so PCR duplication is probably not the culprit.
We can see in the flagstat results that this sample has a large number of total reads, mapped reads and secondary alignment. Therefore, maybe the high number of mapped reads is due to repetitive regions that lead to multiple mapping and more unique reads/higher gene expression. We also see a spike in GC content for many on the experimental groups which might be due to contamination or a biological effect. 

### Rename to know which samples are control and which are BP1
```{}
$ rename SRR8367773 SRR8367773_control SRR8367773*
$ rename SRR8367783 SRR8367783_control SRR8367783*
$ rename SRR8367785 SRR8367785_control SRR8367785*
$ rename SRR8367786 SRR8367786_control SRR8367786*
$ rename SRR8367787 SRR8367787_control SRR8367787*
$ rename SRR8367789 SRR8367789_control SRR8367789*

$ rename SRR8368044 SRR8368044_BP1 SRR8368044*
$ rename SRR8368048 SRR8368048_BP1 SRR8368048*
$ rename SRR8368055 SRR8368055_BP1 SRR8368055*
$ rename SRR8368061 SRR8368061_BP1 SRR8368061*
$ rename SRR8368072 SRR8368072_BP1 SRR8368072*
$ rename SRR8368152 SRR8368152_BP1 SRR8368152*
```

### Generate a read count table
```{}
$ cd $SCRATCH_DIR
$ featureCounts -a hg38.ensGene.gtf -o gene_counts/hg38_fc alignments/SRR8367773_control.Aligned.sortedByCoord.out.bam alignments/SRR8367783_control.Aligned.sortedByCoord.out.bam alignments/SRR8367785_control.Aligned.sortedByCoord.out.bam
alignments/SRR8367786_control.Aligned.sortedByCoord.out.bam
alignments/SRR8367787_control.Aligned.sortedByCoord.out.bam
alignments/SRR8367789_control.Aligned.sortedByCoord.out.bam
alignments/SRR8368044_BP1.Aligned.sortedByCoord.out.bam
alignments/SRR8368048_BP1.Aligned.sortedByCoord.out.bam
alignments/SRR8368055_BP1.Aligned.sortedByCoord.out.bam
alignments/SRR8368061_BP1.Aligned.sortedByCoord.out.bam
alignments/SRR8368072_BP1.Aligned.sortedByCoord.out.bam
alignments/SRR8368152_BP1.Aligned.sortedByCoord.out.bam -s 2 -p --countReadPairs -C
```

parameters: 
-s 1, perform stranded read counting
-p, Specify that input data contain paired-end reads
--countReadPairs, Count read pairs (fragments) instead of reads.
-C, Do not count read pairs that have their two ends mapping to different chromosomes or mapping to same chromosome but on different strands.

### Load read count table
```{r}
read_counts_summary <- read.table('/Users/noelawheeler/Desktop/Analysis of next gen seq data/angsd-project/hg38_fc.summary')
read_counts <- read.table('/Users/noelawheeler/Desktop/Analysis of next gen seq data/angsd-project/hg38_fc', header = TRUE)
```

### Perform the quality control and processing steps
run qorts for quality control
```{}
$ mamba activate qorts
$ sbatch run_qorts.sh $SCRATCH_DIR/alignments
```

```{r}
install.packages("http://hartleys.github.io/QoRTs/QoRTs_LATEST.tar.gz",
repos = NULL,
type="source")
```
```{r}
library(QoRTs)
# make decoder file
incompleteDecoder <- data.frame(unique.ID = c("SRR8367773_control","SRR8367783_control","SRR8367785_control", "SRR8367786_control", "SRR8367787_control", "SRR8367789_control", "SRR8368044_BP1", "SRR8368048_BP1", "SRR8368055_BP1", "SRR8368061_BP1", "SRR8368072_BP1", "SRR8368152_BP1"),
group.ID = c("CONTROL","CONTROL","CONTROL","CONTROL","CONTROL","CONTROL","CASE","CASE","CASE","CASE","CASE","CASE"),  qc.data.dir = c("qort_out/SRR8367773_control","qort_out/SRR8367783_control", "qort_out/SRR8367785_control", "qort_out/SRR8367786_control", "qort_out/SRR8367787_control", "qort_out/SRR8367789_control", "qort_out/SRR8368044_BP1", "qort_out/SRR8368048_BP1", "qort_out/SRR8368055_BP1", "qort_out/SRR8368061_BP1", "qort_out/SRR8368072_BP1", "qort_out/SRR8368052_BP1"));

decoder <- completeAndCheckDecoder(incompleteDecoder)
```

```{r, message=FALSE}
# read in qort results
#res <- read.qc.results.data("/Users/noelawheeler/Desktop/Analysis of next gen seq data/angsd-project", decoder = decoder, calc.DESeq2 = TRUE, calc.edgeR = TRUE)

# generate multi-plot figures (do once)
# makeMultiPlot.all(res,
# outfile.dir = "/Users/noelawheeler/Desktop/Analysis of next gen seq data/angsd-project/qort_out/summaryPlots/",
# plot.device.name = "png")
```
![](/Users/noelawheeler/Desktop/Analysis of next gen seq data/angsd-project/qort_out/summaryPlots/plot-colorByGroup.png)
The Phred quality score looks good for all samples as does gene body coverage. Some issues are that one experimental sample has very high MT content. This might explain why we get really high read counts in that sample for SRR8368044_BP1. We also see that most of the experimental groups have a spike in GC content. This seems strange and might mean some kind of contamination or maybe some biological effect. 

run deSeq2 for processing
```{r, message=FALSE}
library(ggplot2); theme_set(theme_bw(base_size = 16))
library(magrittr)
library(DESeq2)
# fix column names
orig_names <- names(read_counts)
sample_names <- gsub("^alignments\\.|\\.Aligned.*$", "", orig_names)
names(read_counts) <- sample_names
row.names(read_counts) <- make.names(read_counts$Geneid)
# get rid of unnecessary columns
readcounts <- read_counts[ , -c(1:6)]
# get sample infor
sample_info <- data.frame(condition = gsub("SRR[0-9]+_", "", names(readcounts)), row.names = names(readcounts))
sample_info
# make Deseq object
DESeq.ds <- DESeqDataSetFromMatrix(countData = as.matrix(readcounts),
colData = sample_info,
design = ~ condition)
DESeq.ds
```



```{r}
# option to drop sample SRR8368044_BP1 because of mitochondrial contamination
readcounts %<>% dplyr::select(-SRR8368044_BP1)
# create new sample_info object
sample_info <- data.frame(condition = gsub("SRR[0-9]+_", "", names(readcounts)), row.names = names(readcounts))
# new deseq object
DESeq.ds <- DESeqDataSetFromMatrix(countData = as.matrix(readcounts),
colData = sample_info,
design = ~ condition)
DESeq.ds
```


```{r}
val <- colSums(counts(DESeq.ds))
barplot(val, las = 2)
# remove genes with no reads
keep_genes <- rowSums(counts(DESeq.ds)) > 0
DESeq.ds <- DESeq.ds[ keep_genes, ]
```

Normalize for sequence depth
```{r}
# calculate size factors
DESeq.ds <- estimateSizeFactors(DESeq.ds)

## plot size factors
plot( sizeFactors(DESeq.ds), colSums(counts(DESeq.ds)), # assess them
ylab = "library sizes", xlab = "size factors", cex = .6 )
```

```{r}
counts.sf_normalized <- counts(DESeq.ds, normalized=TRUE)
boxplot(counts(DESeq.ds), main = "read counts only", cex = .6, las = 2)
boxplot(counts.sf_normalized, main = "SF normalized", cex = .6, las =2)

```

We can see that a lot of the samples have pretty extreme outliers. Ater normalization these go down somewhat, although we still see very high counts for the one SRR8368061_BP1 largest outlier gene. 

Look at log scaled axis
```{r}
par(mfrow=c(1,2)) # to plot the two box plots next to each other
## bp of non-normalized
boxplot(log2(counts(DESeq.ds) +1), notch=TRUE,
main = "Non-normalized read counts",
ylab ="log2(read counts)", cex = .6, las = 2)
## bp of size-factor normalized values

boxplot(log2(counts(DESeq.ds, normalized=TRUE) +1), notch=TRUE,
        main = "Size-factor-normalized read counts",
        ylab ="log2(read counts)", cex = .6, las = 2,
        mar = c(10, 4, 4, 2), cex.axis = 0.8)
```


When we view these on a log scale the samples look more similarly distributed than before.  
```{r}
# assign the log counts and log norm counts to a distinct matrix within the DESeq.ds object
assay(DESeq.ds, "log.counts") <- log2(counts(DESeq.ds, normalized = FALSE) + 1)
assay(DESeq.ds, "log.norm.counts") <- log2(counts(DESeq.ds, normalized=TRUE) + 1)

# reduce the dependence of the variance
DESeq.rlog <- rlog(DESeq.ds, blind = TRUE)

# plot to view difference in variance
plot(assay(DESeq.ds, "log.norm.counts")[,1:2], cex=.1, main = "size factor and log2-transformed")
plot(assay(DESeq.rlog)[,1:2],
cex=.1, main = "rlog transformed",
xlab = colnames(assay(DESeq.rlog[,1])),
ylab = colnames(assay(DESeq.rlog[,2])) )
```

Sample clustering
```{r}
library(pheatmap)
sampleDists <- dist(t(assay(DESeq.rlog)))
sampleDistMatrix <- as.matrix(sampleDists)
pheatmap(sampleDistMatrix)
```

The heatmaps shows fairly predictable behavior except for the one control sample that clusters more closely with the BP1 samples than the control. We can also see that most of the BP1 samples are more closely correlated with each other than the control samples are. 


```{r}
plotPCA(DESeq.rlog)
```

If we look at the PC1 with 60% of the variance, we can see that most of the BP1 and control samples cluster together except one control sample that clusters close to the BP1 samples. With regards to PC2, we do not really see clustering between samples. 


Perform differential expression analysis
```{r}
# put control samples as baseline
DESeq.ds$condition %<>% relevel(ref="control")
# check that contrast is set up correctly
if (design(DESeq.ds) != ~condition){
  print("We want condition to be fixed effect in DESeq.ds")
}
# final steps in testing differential expression 
DESeq.ds %<>% estimateDispersions()
DESeq.ds %<>% nbinomWaldTest()
DESeq.ds
```

Raw distribution of P-values
```{r}
rowData(DESeq.ds)@listData[["WaldPvalue_condition_BP1_vs_control"]] %>% hist(main="Raw p-values for BP1 vs control")
```

```{r}
DGE.results <- results(DESeq.ds, independentFiltering = TRUE, alpha = 0.05)
head(DGE.results)
summary(DGE.results)
DGE.results$padj %>% hist(breaks=19, main="Adjusted p-values for control vs BP1")
```
We can see that less genes that are now significant which is the expected result.

Find important genes
```{r}
DGE.results.sorted <- DGE.results %>% `[`(order(.$padj),)
head(DGE.results.sorted)
```
```{r}
par(mfrow=c(1,2))
plotCounts(DESeq.ds, gene="ENSG00000198695", normalized = TRUE, xlab="")
plotCounts(DESeq.ds, gene = which.max(DGE.results$padj), xlab="",
main = "Gene with max. p.adj.\n(=least significant)")
```

heatmap of differentially expressed genes
```{r}
library(pheatmap)
# identify genes with the desired adjusted p-value cut-off
DGEgenes <- rownames(subset(DGE.results.sorted, padj < 0.05))
rlog.dge <- DESeq.rlog[DGEgenes,] %>% assay
pheatmap(rlog.dge, scale="row",
show_rownames=FALSE, main="DGE (row-based z-score)")
```

```{r}
# BiocManager::install("EnhancedVolcano")
library(EnhancedVolcano)
vp1 <- EnhancedVolcano(DGE.results,
lab=rownames(DGE.results),
x='log2FoldChange', y='padj',
xlim = c(-10,10),
pCutoff=0.05,
title="SNF2 / WT")
#print(vp1)
```

Shrink logFC values of low/noisily expressed genes
```{r, fig.height=10}
# install.packages("patchwork")
# BiocManager::install("apeglm")
library(patchwork)
DGE.results.shrink <- lfcShrink(DESeq.ds, coef=2, type="apeglm")
vp2 <- EnhancedVolcano(DGE.results.shrink, lab = rownames(DGE.results.shrink), x = "log2FoldChange", y='padj', title="with logFC shrinkage")
vp1 + vp2
```

```{r}
#BiocManager::install("org.Hs.eg.db")
library(org.Hs.eg.db)
human <- org.Hs.eg.db
# gene_symbols <- mapIds(org.Hs.eg.db, keys=ensembl_ids, column="SYMBOL", keytype="ENSEMBL")
annot.DGE <- select(human, keys=rownames(DGE.results.shrink),
keytype="ENSEMBL", columns="SYMBOL")
head(annot.DGE)
```

Find names of most differentially expressed genes
```{r}
# add ENSEMBL as column instead of rowname
DGE.results.sorted$ENSEMBL <- rownames(DGE.results.sorted)
rownames(DGE.results.sorted) <- NULL
# merge annotation and DGE results on common column ENSEMBL
DGE.results.annot <- merge(as.data.frame(DGE.results.sorted), annot.DGE, by = "ENSEMBL", all = TRUE)
# sort results by p-values
DGE.annot.sorted <- DGE.results.annot %>% `[`(order(.$padj),)
head(DGE.annot.sorted)
```
The most important differentially expressed genes are NADH dehydrogenase subunit 6, small nucleolar RNA, H/ACA box 53, NADH dehydrogenase subunit 5, and calcium voltage-gated channel subunit alpha1 E. A lot of these are mitochondrial genes.


Find genes that are highly expressed in the SRR8368152_BP1 sample because it has so many more counts than the other samples.
```{r}
counts_bp1 <- counts(DESeq.ds)[, "SRR8368152_BP1"]
df_bp1 <- as.data.frame(counts_bp1)
df_bp1$ENSEMBL <- rownames(df_bp1)
rownames(df_bp1) <- NULL
# merge annotation and DGE results on common column ENSEMBL
annot.onesample <- merge(df_bp1, annot.DGE, by = "ENSEMBL", all = TRUE)
onesample.counts <- annot.onesample[order(-annot.onesample[,2]),]
head(onesample.counts)
```
```{r}
# run gsea
# get just gene symbol and log fold change 
#BiocManager::install("fgsea")
library(fgsea)
# take only symbol and log fold change from deg result
# take average of duplicate genes
res <- DGE.annot.sorted %>% dplyr::select(SYMBOL, log2FoldChange) %>% 
  na.omit() %>% dplyr::group_by(SYMBOL) %>% dplyr::summarize(log2FoldChange=mean(log2FoldChange))

# turn list into dataframe
ranks <- tibble::deframe(res)
# head(ranks, 20)

# import pathways file to use
pathways.hallmark <- gmtPathways("/Users/noelawheeler/Desktop/Analysis of next gen seq data/angsd-project/h.all.v2023.1.Hs.symbols.gmt")
# check pathways look fine 
pathways.hallmark %>% head() %>% lapply(head)

# run gsea
fgsea.res <- fgseaMultilevel(pathways=pathways.hallmark, stats=ranks)
# select certain columns and rank by p value
fgsea.res.order <- fgsea.res %>% dplyr::select(-log2err, -ES, -leadingEdge, ) %>% dplyr::arrange(padj)
head(fgsea.res.order)
```

```{r}
# view results of gsea normalized enrichment score
library(ggplot2)
# only look at pathways with NES > 0 and NES != na
visualize.fgsea <- fgsea.res.order %>% na.omit(.)

# plot
head(visualize.fgsea)
ggplot(visualize.fgsea, aes(reorder(pathway, NES), NES)) +
geom_col(aes(fill=padj<0.05)) + coord_flip() + labs(x="Pathway", y="Normalized Enrichment Score", title="Hallmark pathways from GSEA") + theme(text = element_text(size = 6))
```
```{r}
# try kegg pathway
pathways.kegg <- gmtPathways("/Users/noelawheeler/Desktop/Analysis of next gen seq data/angsd-project/c2.cp.kegg.v2023.1.Hs.symbols.gmt")
# check pathways look fine 
pathways.kegg %>% head() %>% lapply(head)

# run gsea
fgsea.kegg <- fgseaMultilevel(pathways=pathways.kegg, stats=ranks)
# select certain columns and rank by p value
fgsea.res.order <- fgsea.kegg %>% dplyr::select(-log2err, -ES, -leadingEdge, ) %>% dplyr::arrange(padj)
head(fgsea.res.order)
# remove na
visualize.kegg <- fgsea.res.order %>% na.omit(.)

# plot
ggplot(visualize.kegg, aes(reorder(pathway, NES), NES)) +
geom_col(aes(fill=padj<0.05)) + coord_flip() + labs(x="Pathway", y="Normalized Enrichment Score", title="Hallmark pathways from GSEA") + theme(text = element_text(size = 3))
```
```{r}
visualize.kegg
```
```{r}
plotEnrichment(pathways.kegg[["KEGG_NEUROACTIVE_LIGAND_RECEPTOR_INTERACTION"]], ranks) + labs(title="Neuroactive ligand receptor pathway")
```


